File: elevator-system/go-without-multithreading/elevator.go
package main

import (
	"fmt"
)

// Elevator represents a single physical elevator in the system
type Elevator struct {
	ID               string
	CurrentFloor     int
	CurrentDirection Direction
	Capacity         int
	CurrentLoad      int

	UpStops   map[int]bool // Floors where the elevator needs to stop while moving UP
	DownStops map[int]bool // Floors where the elevator needs to stop while moving DOWN

	Requests []*Request // List of currently assigned requests
}

// NewElevator creates a new elevator initialized to the ground floor (0) and state IDLE
func NewElevator(id string, capacity int) *Elevator {
	return &Elevator{
		ID:               id,
		CurrentFloor:     0,
		CurrentDirection: IDLE,
		Capacity:         capacity,
		CurrentLoad:      0,
		UpStops:          make(map[int]bool),
		DownStops:        make(map[int]bool),
		Requests:         make([]*Request, 0),
	}
}

// AddRequest assigns a new passenger request to this elevator and updates its planned stops
func (e *Elevator) AddRequest(req *Request) error {
	if e.CurrentLoad+req.Passengers > e.Capacity {
		return fmt.Errorf("elevator %s capacity exceeded", e.ID)
	}

	e.Requests = append(e.Requests, req)
	e.CurrentLoad += req.Passengers

	if req.SourceFloor > e.CurrentFloor {
		e.UpStops[req.SourceFloor] = true
	} else if req.SourceFloor < e.CurrentFloor {
		e.DownStops[req.SourceFloor] = true
	} else {
		if req.DestinationFloor > e.CurrentFloor {
			e.UpStops[req.DestinationFloor] = true
		} else {
			e.DownStops[req.DestinationFloor] = true
		}
	}

	if req.DestinationFloor > req.SourceFloor {
		e.UpStops[req.DestinationFloor] = true
	} else if req.DestinationFloor < req.SourceFloor {
		e.DownStops[req.DestinationFloor] = true
	}

	if e.CurrentDirection == IDLE {
		if req.SourceFloor > e.CurrentFloor || req.DestinationFloor > e.CurrentFloor {
			e.CurrentDirection = UP
		} else if req.SourceFloor < e.CurrentFloor || req.DestinationFloor < e.CurrentFloor {
			e.CurrentDirection = DOWN
		}
	}

	return nil
}

// GetDistanceIfAssigned calculates a heuristic "distance" cost for this elevator to handle the given request
// Returns int(1e9) if the elevator is currently too full
func (e *Elevator) GetDistanceIfAssigned(req *Request) int {
	if e.CurrentLoad+req.Passengers > e.Capacity {
		return int(1e9)
	}

	dist := e.CurrentFloor - req.SourceFloor
	if dist < 0 {
		dist = -dist
	}
	
	if e.CurrentDirection == IDLE {
		return dist
	}
	if e.CurrentDirection == UP && req.SourceFloor >= e.CurrentFloor {
		return dist
	}
	if e.CurrentDirection == DOWN && req.SourceFloor <= e.CurrentFloor {
		return dist
	}

	// Penalizes the elevator with a higher cost if moving in the opposite direction
	return dist + 1000 
}

// processCurrentFloor handles stopping and dropping off passengers at the current floor
func (e *Elevator) processCurrentFloor() {
	if e.CurrentDirection == UP {
		if e.UpStops[e.CurrentFloor] {
			fmt.Printf("   -> Elevator %s 🛑 STOPPED at floor %d (UP)\n", e.ID, e.CurrentFloor)
			delete(e.UpStops, e.CurrentFloor)
			e.dropOffPassengers()
		}
	} else if e.CurrentDirection == DOWN {
		if e.DownStops[e.CurrentFloor] {
			fmt.Printf("   -> Elevator %s 🛑 STOPPED at floor %d (DOWN)\n", e.ID, e.CurrentFloor)
			delete(e.DownStops, e.CurrentFloor)
			e.dropOffPassengers()
		}
	}
}

// dropOffPassengers removes completed requests and updates the elevator's current load
func (e *Elevator) dropOffPassengers() {
	var remaining []*Request
	for _, req := range e.Requests {
		if req.DestinationFloor == e.CurrentFloor {
			e.CurrentLoad -= req.Passengers
			fmt.Printf("      -> %d passengers ALIGHTED at floor %d from elevator %s\n", req.Passengers, e.CurrentFloor, e.ID)
		} else {
			remaining = append(remaining, req)
		}
	}
	e.Requests = remaining
}

// getNextDirection determines whether the elevator should continue in its current direction, reverse, or turn IDLE
func (e *Elevator) getNextDirection() Direction {
	if e.CurrentDirection == UP {
		for floor := range e.UpStops {
			if floor > e.CurrentFloor {
				return UP
			}
		}
		if len(e.DownStops) > 0 {
			return DOWN
		}
	} else if e.CurrentDirection == DOWN {
		for floor := range e.DownStops {
			if floor < e.CurrentFloor {
				return DOWN
			}
		}
		if len(e.UpStops) > 0 {
			return UP
		}
	}

	if len(e.UpStops) > 0 {
		return UP
	} else if len(e.DownStops) > 0 {
		return DOWN
	}

	return IDLE
}

// Tick manually steps the exact sequence once per cycle without background threads
// It evaluates direction, moves by one floor, and processes any stops at the new floor
func (e *Elevator) Tick() {
	nextDir := e.getNextDirection()
	e.CurrentDirection = nextDir

	if e.CurrentDirection == IDLE {
		return
	}

	if e.CurrentDirection == UP {
		e.CurrentFloor++
	} else if e.CurrentDirection == DOWN {
		e.CurrentFloor--
	}

	e.processCurrentFloor()
}

File: elevator-system/java-without-multithreading/Elevator.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Represents a single elevator, managing its state and processing requests.
 */
public class Elevator {
    private final String id;
    private int currentFloor;
    private Direction currentDirection;
    private final int capacity;
    private int currentLoad;

    private final Map<Integer, Boolean> upStops;
    private final Map<Integer, Boolean> downStops;
    private List<Request> requests;

    public Elevator(String id, int capacity) {
        this.id = id;
        this.currentFloor = 0;
        this.currentDirection = Direction.IDLE;
        this.capacity = capacity;
        this.currentLoad = 0;
        this.upStops = new HashMap<>();
        this.downStops = new HashMap<>();
        this.requests = new ArrayList<>();
    }

    public String getId() {
        return id;
    }

    public Direction getCurrentDirection() {
        return currentDirection;
    }

    /**
     * Adds a request to this elevator if capacity allows, updating its internal stops.
     */
    public void addRequest(Request req) throws Exception {
        if (currentLoad + req.getPassengers() > capacity) {
            throw new Exception("elevator " + id + " capacity exceeded");
        }

        requests.add(req);
        currentLoad += req.getPassengers();

        if (req.getSourceFloor() > currentFloor) {
            upStops.put(req.getSourceFloor(), true);
        } else if (req.getSourceFloor() < currentFloor) {
            downStops.put(req.getSourceFloor(), true);
        } else {
            if (req.getDestinationFloor() > currentFloor) {
                upStops.put(req.getDestinationFloor(), true);
            } else {
                downStops.put(req.getDestinationFloor(), true);
            }
        }

        if (req.getDestinationFloor() > req.getSourceFloor()) {
            upStops.put(req.getDestinationFloor(), true);
        } else if (req.getDestinationFloor() < req.getSourceFloor()) {
            downStops.put(req.getDestinationFloor(), true);
        }

        if (currentDirection == Direction.IDLE) {
            if (req.getSourceFloor() > currentFloor || req.getDestinationFloor() > currentFloor) {
                currentDirection = Direction.UP;
            } else if (req.getSourceFloor() < currentFloor || req.getDestinationFloor() < currentFloor) {
                currentDirection = Direction.DOWN;
            }
        }
    }

    /**
     * Calculates the estimated distance/cost for an elevator to serve a new request.
     * @return An integer representing distance, or Integer.MAX_VALUE if capacity is exceeded.
     */
    public int getDistanceIfAssigned(Request req) {
        if (currentLoad + req.getPassengers() > capacity) {
            return Integer.MAX_VALUE;
        }

        int dist = Math.abs(currentFloor - req.getSourceFloor());

        if (currentDirection == Direction.IDLE) {
            return dist;
        }
        if (currentDirection == Direction.UP && req.getSourceFloor() >= currentFloor) {
            return dist;
        }
        if (currentDirection == Direction.DOWN && req.getSourceFloor() <= currentFloor) {
            return dist;
        }

        return dist + 1000;
    }

    /**
     * Processes stops at the current floor, dropping off passengers and removing fulfilled stops.
     */
    private void processCurrentFloor() {
        if (currentDirection == Direction.UP) {
            if (upStops.containsKey(currentFloor) && upStops.get(currentFloor)) {
                System.out.printf("   -> Elevator %s \uD83D\uDED1 STOPPED at floor %d (UP)\n", id, currentFloor);
                upStops.remove(currentFloor);
                dropOffPassengers();
            }
        } else if (currentDirection == Direction.DOWN) {
            if (downStops.containsKey(currentFloor) && downStops.get(currentFloor)) {
                System.out.printf("   -> Elevator %s \uD83D\uDED1 STOPPED at floor %d (DOWN)\n", id, currentFloor);
                downStops.remove(currentFloor);
                dropOffPassengers();
            }
        }
    }

    private void dropOffPassengers() {
        List<Request> remaining = new ArrayList<>();
        for (Request req : requests) {
            if (req.getDestinationFloor() == currentFloor) {
                currentLoad -= req.getPassengers();
                System.out.printf("      -> %d passengers ALIGHTED at floor %d from elevator %s\n", 
                        req.getPassengers(), currentFloor, id);
            } else {
                remaining.add(req);
            }
        }
        requests = remaining;
    }

    /**
     * Determines the next direction for the elevator based on remaining requests.
     */
    private Direction getNextDirection() {
        if (currentDirection == Direction.UP) {
            for (int floor : upStops.keySet()) {
                if (floor > currentFloor) {
                    return Direction.UP;
                }
            }
            if (!downStops.isEmpty()) {
                return Direction.DOWN;
            }
        } else if (currentDirection == Direction.DOWN) {
            for (int floor : downStops.keySet()) {
                if (floor < currentFloor) {
                    return Direction.DOWN;
                }
            }
            if (!upStops.isEmpty()) {
                return Direction.UP;
            }
        }

        if (!upStops.isEmpty()) {
            return Direction.UP;
        } else if (!downStops.isEmpty()) {
            return Direction.DOWN;
        }

        return Direction.IDLE;
    }

    /**
     * Simulates one unit of time, moving the elevator and processing stops.
     */
    public void tick() {
        Direction nextDir = getNextDirection();
        currentDirection = nextDir;

        if (currentDirection == Direction.IDLE) {
            return;
        }

        if (currentDirection == Direction.UP) {
            currentFloor++;
        } else if (currentDirection == Direction.DOWN) {
            currentFloor--;
        }

        processCurrentFloor();
    }
}
